### DISCUSSION TITLE
Found some strange dart interface

### DISCUSSION THREAD

**Discussion by HaloWang** [halildurmus/dartwinrt#383] [Q&A] [answered]
I'm coping some cpp codes into dart

this is the source cpp codes:

```c++
BYTE* lp = r.lpbitmap;

InMemoryRandomAccessStream stream;
DataWriter dataWriter{stream};

dataWriter.WriteBytes(array_view<byte>(lp, r.dwBmpSize));
auto buffer = dataWriter.DetachBuffer();
```

this is what I write using dartwinrt

```dart
final List<int> bytes = ...;
final stream = InMemoryRandomAccessStream();
final DataWriter dataWriter = DataWriter.createDataWriter(stream);
dataWriter.writeBytes(bytes);
final buffer = dataWriter.detachBuffer();
```

I wonder could I use something like `Pointer<List<int>>` to write to dataWriter on the dart-side?

And why `DataWriter` in dart hasn't method matching `WriteBytes(array_view<uint8_t const> value)` on the cpp-side?

**Comment by halildurmus** [comment, marked as answer]
> ...
> I wonder could I use something like `Pointer<List<int>>` to write to dataWriter on the dart-side?

No, you can't (and you likely meant `Pointer<Uint8>` instead of `Pointer<List<int>>`, as the latter is not a valid Dart type).

> And why `DataWriter` in dart hasn't method matching `WriteBytes(array_view<uint8_t const> value)` on the cpp-side?

This project is designed to provide an idiomatic Dart projection of the WinRT APIs, which involves substantial wrapping. For the sake of idiomatic usage, the method doesn't take a parameter like `Pointer<Uint8>`; instead, it accepts `List<int>`, which serves as the equivalent counterpart.

**Comment by HaloWang** [comment]
OK, and I find I can directly use `Pointer<Uint8>.asTypedList` provided by dart:ffi.

This function can easily convert `Pointer<Uint8>` to `Uint8List`, which implements `List<int>`

```dart
final dwBmpSize = 1000;
final Pointer<Uint8> lpBitmap = arena<Uint8>(dwBmpSize);
final Uint8List bytes = lpBitmap.asTypedList(dwBmpSize);
// ...
dataWriter.writeBytes(bytes);
```

**Reply by halildurmus** [reply]
Nice!
